### ISSUE TITLE
How to recovery the step count of scheduler when recovery a training?

### ISSUE THREAD

**Issue by chufall** [google-deepmind/optax#1036]
Hi, I  have a question about the recoverying the step count  of scheduler。

The example code is the following:
````
import jax
import optax
import jax.numpy as jnp
import functools

@functools.partial(jax.vmap, in_axes=(None, 0))
def network(params, x):
  return jnp.dot(params, x)

def compute_loss(params, x, y):
  y_pred = network(params, x)
  loss = jnp.mean(optax.l2_loss(y_pred, y))
  return loss

key = jax.random.PRNGKey(42)
target_params = 0.5

# Generate some data.
xs = jax.random.normal(key, (16, 2))
ys = jnp.sum(xs * target_params, axis=-1)

# Exponential decay of the learning rate.
scheduler = optax.exponential_decay(
    init_value=0.01,
    transition_steps=1000,
    decay_rate=0.99)

# Combining gradient transforms using optax.chain.
optimizer = optax.chain(
    optax.scale_by_schedule(scheduler),
    optax.scale_by_adam() 
)

# Initialize parameters of the model + optimizer.
params = jnp.array([0.0, 0.0])  # Recall target_params=0.5.
#gradient_transform_1 = optax.inject_hyperparams(gradient_transform)
opt_state = optimizer.init(params)  #初始化后，optimizer的scheduler.count总是从0开始
print(f"Init learning rate : {scheduler(opt_state[0].count)} , and step count is {opt_state[0].count}")

# A simple update loop.
for _ in range(10):
  grads = jax.grad(compute_loss)(params, xs, ys)
  updates, opt_state = optimizer.update(grads, opt_state)
  print(f"current learning rate : {scheduler(opt_state[0].count)}, and step count is {opt_state[0].count}")
  params = optax.apply_updates(params, updates)

# **Simulating when the training is broken, and recovery the training from the existed params**
print("\nSimulating when the training is broken, and recovery the training from the existed params")
opt_state = optimizer.init(params)  
print(f"Loaded learning rate : {scheduler(opt_state[0].count)}, and step count is {opt_state[0].count}")

# update loop.
for _ in range(2):
  grads = jax.grad(compute_loss)(params, xs, ys)
  updates, opt_state = optimizer.update(grads, opt_state)
  print(f"current learning rate : {scheduler(opt_state[0].count)}, and step count is {opt_state[0].count}")
  params = optax.apply_updates(params, updates)
````

The running result is as bellow ：
```
Init learning rate : 0.009999999776482582 , and step count is 0
current learning rate : 0.009999899193644524, and step count is 1
current learning rate : 0.009999798610806465, and step count is 2
current learning rate : 0.009999698027968407, and step count is 3
current learning rate : 0.009999598376452923, and step count is 4
current learning rate : 0.00999949686229229, and step count is 5
current learning rate : 0.009999396279454231, and step count is 6
current learning rate : 0.009999296627938747, and step count is 7
current learning rate : 0.009999196045100689, and step count is 8
current learning rate : 0.00999909546226263, and step count is 9
current learning rate : 0.009998994879424572, and step count is 10

Simulating when the training is broken, and recovery the training from the existed params
Loaded learning rate : 0.009999999776482582, and step count is 0
current learning rate : 0.009999899193644524, and step count is 1
current learning rate : 0.009999798610806465, and step count is 2
```

I found the step count of scheduler did always start from 0 and not from 10 after recovery the traning .

I can record the step count, and  there's any way to set the step count into the scheduler to recovery the training from the correct learning rate.

Thanks a lot!

Qc

**Comment by fabianp** [comment]
hey, I think just commenting out the second `opt_state = optimizer.init(params)` should do the trick

**Comment by chufall** [comment]
> hey, I think just commenting out the second `opt_state = optimizer.init(params)` should do the trick

Thank you for reply!

I have to say you have misunderstand my code.
The second opt_state = optimizer.init(params) is used to simulate:
 When one training is broken , I  have to  use the last params to init the optimizer and continue the training.
But in this situation,  I cannot set the step count into the optimizer and let the schedule continue the step rather than start from  zero again.  
Thanks a lot!

**Comment by vroulet** [comment]
Hello @chufall,
You need to have saved both params and state. The count just cannot be inferred from the parameters only. If the training is broken you need to have saved not only the params but also the optimizer state, that's the only way you can keep accessing the count. (Note that when using optimizers like adam that keep in memory some momentums in their states you also need to have saved somewhere these momentums when you want to "restart" the optimization. Otherwise the momentums are reinitialized at zero and lots of bad things could happen).

**Comment by chufall** [comment]
> Hello @chufall, You need to have saved both params and state. The count just cannot be inferred from the parameters only. If the training is broken you need to have saved not only the params but also the optimizer state, that's the only way you can keep accessing the count. (Note that when using optimizers like adam that keep in memory some momentums in their states you also need to have saved somewhere these momentums when you want to "restart" the optimization. Otherwise the momentums are reinitialized at zero and lots of bad things could happen).

Thank you for your reply!
But how to implement it in the optax?  I have no idea about how to recovery the step count of optax.exponential_decay

**Comment by vroulet** [comment]
I may not understand your issue.
In your training loop, you need to have some lines like (take a look at e.g. orbax to see how checkpointing can be implemented).
```
if (iteration % save_every_iteration) == 0:
  save(params, opt_state, ...)
```
And when you reload from your past checkpoint, 
```
params, opt_state = load(...)
# Do your usual training loop
```
See e.g. https://github.com/google/flax/blob/main/examples/imagenet/train.py for an example of setting up checkpoints and reloading from them.

If you are asking about how to access one particular entry from a state, maybe #961 could be useful.
